pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,354 @@
+	// this VolumeSnapshotContent's name for the bidirectional binding to be valid.
+	// For a pre-existing VolumeSnapshotContent object, name and namespace of the
+	// VolumeSnapshot object MUST be provided for binding to happen.
+	// Required.
+	VolumeSnapshotRef core_v1.ObjectReference `json:"volumeSnapshotRef" protobuf:"bytes,1,opt,name=volumeSnapshotRef"`
+
+	// deletionPolicy determines whether this VolumeSnapshotContent and its physical snapshot on
+	// the underlying storage system should be deleted when its bound VolumeSnapshot is deleted.
+	// Supported values are "Retain" and "Delete".
+	// "Retain" means that the VolumeSnapshotContent and its physical snapshot on underlying storage system are kept.
+	// "Delete" means that the VolumeSnapshotContent and its physical snapshot on underlying storage system are deleted.
+	// In dynamic snapshot creation case, this field will be filled in with the "DeletionPolicy" field defined in the
+	// VolumeSnapshotClass the VolumeSnapshot refers to.
+	// For pre-existing snapshots, users MUST specify this field when creating the VolumeSnapshotContent object.
+	// +kubebuilder:validation:Enum=Delete;Retain
+	// Required.
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy" protobuf:"bytes,2,opt,name=deletionPolicy"`
+
+	// driver is the name of the CSI driver used to create the physical snapshot on
+	// the underlying storage system.
+	// This MUST be the same as the name returned by the CSI GetPluginName() call for
+	// that driver.
+	// Required.
+	Driver string `json:"driver" protobuf:"bytes,3,opt,name=driver"`
+
+	// name of the SnapshotClass to which this snapshot belongs.
+	// +optional
+	SnapshotClassName *string `json:"snapshotClassName,omitempty" protobuf:"bytes,4,opt,name=snapshotClassName"`
+
+	// source specifies from where a snapshot will be created.
+	// Required.
+	Source VolumeSnapshotContentSource `json:"source" protobuf:"bytes,5,opt,name=source"`
+}
+
+// VolumeSnapshotContentSource represents the source of a snapshot.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotContentSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotContentSource members
+// will not be updated once specified.
+type VolumeSnapshotContentSource struct {
+	// volumeHandle specifies the CSI name of the volume from which a snapshot
+	// should be dynamically taken from.
+	// +optional
+	VolumeHandle *string `json:"volumeHandle,omitempty" protobuf:"bytes,1,opt,name=volumeHandle"`
+
+	// snapshotHandle specifies the CSI name of a pre-existing snapshot on the
+	// underlying storage system.
+	// +optional
+	SnapshotHandle *string `json:"snapshotHandle,omitempty" protobuf:"bytes,2,opt,name=snapshotHandle"`
+}
+
+// VolumeSnapshotContentStatus is the status of a VolumeSnapshotContent object
+type VolumeSnapshotContentStatus struct {
+	// snapshotHandle is the CSI name of a snapshot on the underlying storage system.
```

Comment:
that...?

Reply:
What happens when it's not set? Is there a difference between nil and empty string?

Reply:
empty string should be treated the same as any other string. Again, sidecar controller will be the only one updating (unless there are other controllers trying to override this field, thus empty string should not be a possible option.)
if it's nil (not specified), which is only possible in dynamic snapshot creation case, it indicates the snapshot creation either failed or still in progress.

Reply:
I will update the comments to indicate that

Reply:
Refer to the CSI spec names instead of "CSI name of a snapshot".

In pre-provisioned case, is SnapshotHandle also set?